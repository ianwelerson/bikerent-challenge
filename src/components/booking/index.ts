export { default as BookingAddressMap } from './BookingAddressMap.vue'
export { default as BookingPricing } from './BookingPricing.vue'
export { default as BookingOverview } from './BookingOverview.vue'
export { default as BookingSuccess } from './BookingSuccess.vue'
