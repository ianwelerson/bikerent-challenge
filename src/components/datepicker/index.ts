export { default as Datepicker } from './DatepickerInput.vue'
