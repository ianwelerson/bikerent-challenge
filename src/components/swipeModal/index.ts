export { default as SwipeModal } from './SwipeModal.vue'
