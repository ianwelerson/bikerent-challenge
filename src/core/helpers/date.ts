/**
 *
 * @param {string} dateString YYYY-MM-DD date
 * @param {part} part The part you want to retrieve, it can be YYYY, MM or DD
 * @returns {string} required part
 */
export function formatDate(dateString: string, part: 'YYYY' | 'MM' | 'DD'): string | null {
  const dateArray = new Date(dateString).toISOString().substring(0, 10).split('-')

  switch (part) {
    case 'YYYY':
      return dateArray[0]
    case 'MM':
      return dateArray[1]
    case 'DD':
      return dateArray[2]
  }
}

/**
 *
 * @param {number} year the year
 * @param {number} month the month number from 1 to 12
 * @param {number} length the length of the array
 * @returns {string[]} array of dates
 */
export function generateDatepickerDays(year: number, month: number, length: number): string[] {
  const days = []

  const normalizedMonth = month - 1 // Zero based
  const currentMonthStart = new Date(year, normalizedMonth, 1)
  const currentMonthEnd = new Date(year, normalizedMonth + 1, 0)

  const currentStartWeekday = currentMonthStart.getDay()
  const currentEndWeekday = currentMonthEnd.getDay()

  // Previous month
  for (let i = currentStartWeekday - 1; i >= 0; i--) {
    days.push(new Date(year, month - 1, -i).toISOString().split('T')[0])
  }

  // Current
  for (let i = 1; i <= currentMonthEnd.getDate(); i++) {
    days.push(new Date(year, normalizedMonth, i).toISOString().split('T')[0])
  }

  // Next month
  for (let i = 0; i < 6 - currentEndWeekday; i++) {
    days.push(new Date(year, month, i + 1).toISOString().split('T')[0])
  }

  /**
   * The slice ensure the correct size to match with the Figma layout
   * where we have seven columns and five rows.
   * But this introduce a UX problem: for some months we need to have six rows
   */
  return days.slice(0, length)
}
