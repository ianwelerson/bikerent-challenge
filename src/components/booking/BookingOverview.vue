<script lang="ts">
import { defineComponent, type PropType } from 'vue'
import axios from 'axios'

import { CurrencyCode } from '@/core/config'

// import { bike } from '@/core/api'
import type { BikeRentDetails } from '@/core/api/modules/typings/bike'

import { BookingPricing } from '@/components/booking'
import { LoadingSpinner } from '@/components/loading'

export default defineComponent({
  name: 'BookingOverview',
  components: {
    BookingPricing,
    LoadingSpinner
  },
  props: {
    rentDetails: {
      type: Object as PropType<BikeRentDetails>,
      required: true
    }
  },
  emits: ['booked'],
  data: () => ({
    currency: CurrencyCode.EUR,
    isLoading: false,
    error: null as string | null
  }),
  watch: {
    isLoading() {
      this.error = null
    }
  },
  methods: {
    errorHandle(message: string | null) {
      this.error = message
    },
    async handleAddBooking() {
      this.isLoading = true

      try {
        // await bike.rent(this.rentDetails)
        await new Promise((resolve) => {
          setTimeout(() => {
            resolve(null)
          }, 3000)
        })
        this.$emit('booked')
      } catch (error) {
        if (axios.isAxiosError(error)) {
          this.errorHandle(error.response?.data.message)
        }
      } finally {
        this.isLoading = false
      }
    }
  }
})
</script>

<template>
  <div>
    <h3 class="text-base mb-4">Booking Overview</h3>

    <div class="divider" />

    <booking-pricing
      :currency="currency"
      :details="rentDetails"
      class="mb-8"
      @updated="isLoading = false"
      @error="errorHandle"
    />

    <button
      :class="['button button--primary w-full py-5', { 'btn-disabled': isLoading || !!error }]"
      :disabled="isLoading || !!error"
      @click="handleAddBooking"
    >
      <loading-spinner v-if="isLoading" />
      <div v-else-if="!!error" class="error-message">{{ error }}</div>
      <template v-else>Add to booking</template>
    </button>
  </div>
</template>

<style scoped lang="scss">
.error-message {
  color: get-theme-color('error');
}
</style>
