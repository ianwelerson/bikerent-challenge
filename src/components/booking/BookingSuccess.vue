<script lang="ts">
import { defineComponent } from 'vue'

import { Chip } from '@/components/chip'
import { ImageLazy } from '@/components/image'

export default defineComponent({
  name: 'BookingSuccess',
  components: {
    Chip,
    ImageLazy
  },
  props: {
    thumb: {
      type: String,
      required: true
    },
    name: {
      type: String,
      required: true
    },
    type: {
      type: String,
      required: true
    },
    showRedirect: {
      type: Boolean,
      default: false
    }
  }
})
</script>

<template>
  <div class="rent-success">
    <div>
      <h3 class="mb-6 text-2xl font-extrabold">Thank you!</h3>
      <p class="m-0 text-base font-semibold">Your bike is booked.</p>
    </div>
    <div class="mt-6">
      <image-lazy :src="thumb" class="rent-success__img" />
      <div class="mt-4">
        <h2 class="text-lg font-bold">{{ name }}</h2>
        <chip color="secondary" size="sm">{{ type }}</chip>
      </div>
    </div>
    <div v-if="showRedirect" class="mt-5">
      <router-link :to="{ name: 'home' }" class="button button--primary w-full">Go to Home Page</router-link>
    </div>
  </div>
</template>

<style scoped lang="scss">
.rent-success {
  padding: 39px 0;
  @apply .text-center;

  &__img {
    width: 185px;
  }
}
</style>
