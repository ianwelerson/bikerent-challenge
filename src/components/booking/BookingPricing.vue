<script lang="ts">
import { defineComponent, type PropType } from 'vue'
import axios from 'axios'
import { Icon } from '@/components/icon/'
import { formatCurrency } from '@/core/helpers/currency'
import { bike } from '@/core/api'

import type { CurrencyCode } from '@/core/config'
import type { BikeRentDetails } from '@/core/api/modules/typings/bike'

export default defineComponent({
  name: 'BookingPricing',
  components: { Icon },
  props: {
    currency: {
      type: String as PropType<CurrencyCode>,
      required: true
    },
    details: {
      type: Object as PropType<BikeRentDetails>,
      required: true
    }
  },
  emits: ['loading', 'updated', 'error'],
  data: () => ({
    subtotal: 0,
    serviceFee: 0,
    total: 0
  }),
  watch: {
    details() {
      this.updatePrice()
    }
  },
  mounted() {
    this.updatePrice()
  },
  methods: {
    formatPrice(value: number): String {
      return formatCurrency(value, this.currency)
    },
    async updatePrice() {
      this.$emit('loading')

      try {
        const { rentAmount, fee, totalAmount } = await bike.amount(this.details)

        this.subtotal = rentAmount
        this.serviceFee = fee
        this.total = totalAmount
      } catch (error) {
        if (axios.isAxiosError(error)) {
          this.$emit('error', error.response?.data.message)
        }
      } finally {
        this.$emit('updated')
      }
    }
  }
})
</script>

<template>
  <section class="booking-pricing">
    <div class="grid grid-cols-2 gap-y-3 text-sm">
      <div class="flex items-center">
        Subtotal
        <icon type="solid" size="xs" class="ml-1">circle-info</icon>
      </div>
      <div class="text-right">{{ formatPrice(subtotal) }}</div>

      <div class="flex items-center">
        Service Fee
        <icon type="solid" size="xs" class="ml-1">circle-info</icon>
      </div>
      <div class="text-right">{{ formatPrice(serviceFee) }}</div>

      <div class="flex items-center text-base">Total</div>
      <div class="text-right text-2xl font-bold">{{ formatPrice(total) }}</div>
    </div>
  </section>
</template>

<style lang="scss" scoped>
.booking-pricing {
  position: relative;
}
</style>
