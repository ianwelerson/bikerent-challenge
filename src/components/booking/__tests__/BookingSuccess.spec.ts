import { describe, expect, it, vi } from 'vitest'
import { mount } from '@vue/test-utils'

import { mockIntersectionObserver } from '@/core/helpers/tests'
import { mockedBike } from '@/mocks'

import { BookingSuccess } from '@/components/booking'

const BIKE_DATA = {
  thumb: mockedBike.cardImage!,
  name: mockedBike.name,
  type: mockedBike.type
}

describe('BookingSuccess', () => {
  vi.stubGlobal('IntersectionObserver', mockIntersectionObserver)

  it('renders correctly', () => {
    const wrapper = mount(BookingSuccess, {
      props: { ...BIKE_DATA }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should hide the redirect button when showRedirect is false', () => {
    const wrapper = mount(BookingSuccess, {
      props: { ...BIKE_DATA }
    })

    expect(wrapper.find('.button--primary').exists()).toBeFalsy()
  })

  it('should show the redirect button', async () => {
    const wrapper = mount(BookingSuccess, {
      props: {
        ...BIKE_DATA,
        showRedirect: true
      }
    })

    expect(wrapper.find('.button--primary').exists()).toBeTruthy()
  })
})
