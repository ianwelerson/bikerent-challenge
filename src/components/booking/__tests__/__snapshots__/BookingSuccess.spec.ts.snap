// Vitest Snapshot v1

exports[`BookingSuccess > renders correctly 1`] = `
"<div class=\\"rent-success\\" data-v-65a8440a=\\"\\">
  <div data-v-65a8440a=\\"\\">
    <h3 class=\\"mb-6 text-2xl font-extrabold\\" data-v-65a8440a=\\"\\">Thank you!</h3>
    <p class=\\"m-0 text-base font-semibold\\" data-v-65a8440a=\\"\\">Your bike is booked.</p>
  </div>
  <div class=\\"mt-6\\" data-v-65a8440a=\\"\\"><img class=\\"image image--lazy rent-success__img\\" data-v-ff3f7011=\\"\\" data-v-65a8440a=\\"\\">
    <div class=\\"mt-4\\" data-v-65a8440a=\\"\\">
      <h2 class=\\"text-lg font-bold\\" data-v-65a8440a=\\"\\">Murazik, Thiel and Robel</h2>
      <div class=\\"chip chip--sm chip--secondary\\" data-v-221b4c58=\\"\\" data-v-65a8440a=\\"\\">
        <div class=\\"chip__content\\" data-v-221b4c58=\\"\\">Cyclocross Bicycle</div>
      </div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
